R/indentation_linter.R

```diff
@@ -0,0 +1,163 @@
+#' Check that indentation is consistent
+#'
+#' @param indent Block indentation level, used for multi-line code blocks (`{ ... }`), function calls (`( ... )`) and
+#'   extractions (`[ ... ]`, `[[ ... ]]`).
+#' @param use_hybrid_indent Require a block indent for multi-line function calls if there are only unnamed arguments
+#'   in the first line?
+#'   ```r
+#'   # complies only with use_hybrid_indent = TRUE:
+#'   map(x, f,
+#'     additional_arg = 42
+#'   )
+#'
+#'   # complies only with use_hybrid_indent = FALSE:
+#'   map(x, f,
+#'       additional_arg = 42
+#'   )
+#'
+#'   # always complies
```

Comment:
causes me some "alarm" that {styler} will edit this style:

```r
styler::style_text("
map(x, f,
    additional_arg = 42)
")

# map(x, f,
#   additional_arg = 42
# )
```

Is totally addressing this just outside the scope of this linter? or should the defaults for this linter move the user towards the {styler} output?

The style guide section is not _quite_ conclusive here, but collating a few BAD examples I think we can conclude {styler} is doing the "right" thing per style guide:

https://style.tidyverse.org/syntax.html#long-lines

Reply:
Maybe we need `function_right_parentheses_linter()`, or just combine into a single `function_parentheses_linter()`?

Reply:
Hmm I think we should at least add a parameter to allow suppressing this lint, because it will be produced by RStudios and PyCharms auto-indent.
Alternatively, we control the expected indentation of function calls based on whether the closing parenthesis has arguments on its line? That would let both styles, hanging indent and ggplot-like indent, pass.
I think the part of the style guide that says function arguments on multiline function calls need to be unnamed if on the first line and named if on subsequent lines should be put into its own linter, though.

Reply:
Oh wait, I think that's what I did with `use_hybrid_indent = TRUE` 😆

Reply:
WDYT about linting

```
map(x, f,
    additional_arg = 42)
```

With a multiline function call linter that also checks that each argument on the subsequent lines is named and the arguments on the same line as the function call are unnamed?
Could be flagged as "Closing parenteses of multiline function calls should always be on their own line".

Reply:
#1719